### PR TITLE
Add source encoding parameter in Memory module

### PR DESCRIPTION
**Pull request by ryanzz**
When caching a function's return value,  current implementation assume that the source file encoding of that function is the system's default encoding. If not, something bad happens.
This pull request added a source code encoding parameter to fix this problem. I think it would be useful.

### CODE REVIEW

**Comment by GaelVaroquaux** [comment]
Could you add a test,  please (one that fails without your patch). You will probably need to add a file with a special encoding.

**Comment by ryanzz** [comment]
OK. I will write a test for it.

**Comment by lesteve** [comment]
> OK. I will write a test for it.

Also, just in case you missed it the Travis tests fail for python 2.6 and 2.7.

**Comment by ryanzz** [comment]
> Also, just in case you missed it, the Travis tests fail for python 2.6 and 2.7.

@lesteve thank you, I have noticed that. And I probably know why.

**Comment by GaelVaroquaux** [comment]
> OK. I will write a test for it.

Thank you!

File: joblib/func_inspect.py [hunk 1]
```diff
@@ -80,6 +88,7 @@ def _clean_win_chars(string):
         quote = urllib.quote
     else:
         # In Python 3, quote is elsewhere
+        import urllib.parse
```
Comment (ogrisel) [inline joblib/func_inspect.py]: why is this needed?

**Comment by ogrisel** [comment]
@ryanzz would you feel comfortable in squashing all those commits (without the merge commits) into a single commit rebased on top of the current master? That would help use keep a more linear git history without any intermediary debug commits.

**Comment by ogrisel** [comment]
I also wonder if there is no better way to ask the Python runtime / interpreter to find if a given Python source files needs to be open with a specific file encoding in `get_func_code` (instead of having to manually pass the right encoding explicitly).

**Comment by ogrisel** [comment]
After a bit of googling I did not find anything convincing. Maybe we could open with `ascii` `error='ignore'` by default and look for the `# -*- coding: encoding_name -*-` pattern with a regexp and reopen the source file with the right encoding if the pattern is found. WDYT?

File: joblib/func_inspect.py (hunk 1 above)
Comment (ryanzz) [inline joblib/func_inspect.py]: In python3.
```
>>> import urllib
>>> urllib.parse
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'parse'
```

Comment (ogrisel) [inline joblib/func_inspect.py]: Indeed.

**Comment by ryanzz** [comment]
@ogrisel 
>would you feel comfortable in squashing all those commits (without the merge commits) into a single commit rebased on top of the current master? 

I am glad to do it. And I also think it would be great if the source code encoding can be detected in the runtime.  You solution is probaly feasible, and really more nonintrusive to solve this problem.

**Comment by ogrisel** [comment]
> And I also think it would be great if the source code encoding can be detected in the runtime. Your solution is probaly feasible, and really more nonintrusive to solve this problem.

Good. Let us know via a comment if you can make it work. (just pushing new commits does not trigger new notifications).

**Comment by ryanzz** [comment]
Hi, I just refactored the code. In the get_func_code, it can open the python source file in the right way now.

**Comment by ogrisel** [comment]
This looks good to me, thanks @ryanzz!

File: joblib/memory.py [hunk 2]
```diff
@@ -540,8 +542,8 @@ def _write_func_code(self, filename, func_code, first_line):
         # sometimes have several functions named the same way in a
         # file. This is bad practice, but joblib should be robust to bad
         # practice.
-        func_code = '%s %i\n%s' % (FIRST_LINE_TEXT, first_line, func_code)
-        with open(filename, 'w') as out:
+        func_code = u'%s %i\n%s' % (FIRST_LINE_TEXT, first_line, func_code)
+        with open(filename, 'w', encoding="UTF-8") as out:
```
Comment (lesteve) [inline joblib/memory.py]: Maybe it's just me but I got a bit confused how this line could even work for python 2.7 since open doesn't have a 'encoding' parameter.

Is it not preferrable to do `import io` and use `io.open` here rather than `from io import open`?

File: joblib/test/test_func_inspect.py [hunk 3]
```diff
@@ -202,3 +202,23 @@ def test_format_signature_numpy():
     """ Test the format signature formatting with numpy.
     """
 
+
+def test_special_source_encoding():
+    from .test_func_inspect_special_encoding import big5_f
+    func_code, source_file, first_line = get_func_code(big5_f)
+    nose.tools.assert_equal(first_line, 5)
+    nose.tools.assert_equal("def big5_f():" in func_code, True)
+    nose.tools.assert_equal("test_func_inspect_special_encoding" in source_file, True)
```
Comment (lesteve) [inline joblib/test/test_func_inspect.py]: you can use assert_true in this line and the line above

File: joblib/memory.py (hunk 2 above)
Comment (ryanzz) [inline joblib/memory.py]: You are right, using io.open is better.

File: joblib/test/test_func_inspect.py (hunk 3 above)
Comment (ryanzz) [inline joblib/test/test_func_inspect.py]: @lesteve thank you,  I will revise it.

**Comment by lesteve** [comment]
Also small detail for your next PR: it's considered best practice to work in a separate branch than master. See for example [this](http://scikit-learn.org/stable/developers/#how-to-contribute) for more details.

File: joblib/func_inspect.py (hunk 1 above)
Comment (lesteve) [inline joblib/func_inspect.py]: Just to share the WTF moment I just had, the same two lines of code work fine inside IPython:

```
Python 3.4.3 |Continuum Analytics, Inc.| (default, Mar  6 2015, 12:03:53) 
Type "copyright", "credits" or "license" for more information.

```
IPython 3.1.0 -- An enhanced Interactive Python.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://binstar.org
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import urllib

In [2]: urllib.parse
Out[2]: <module 'urllib.parse' from '/volatile/le243287/miniconda3/lib/python3.4/urllib/parse.py'>
```

I am guessing somewhere in the IPython code, `urllib.parse` is imported ...

**Comment by lesteve** [comment]
Merging this one, thanks a lot @ryanzz !
